tools/topology/sof-glk-da7219.m4

```diff
@@ -190,6 +192,37 @@ DAI_CONFIG(HDA, 4, 4, iDisp2,
 	HDA_CONFIG(HDA_CONFIG_DATA(HDA, 4, 48000, 2)))
 DAI_CONFIG(HDA, 5, 5, iDisp3,
 	HDA_CONFIG(HDA_CONFIG_DATA(HDA, 5, 48000, 2)))
+', HEADPHONE, `cs42l42', `
+#SSP 2 (ID: 0) with 19.2 MHz mclk with MCLK_ID 1 (unused), 3.072 MHz bclk, no quirk, 10 ms BCLK delay
+DAI_CONFIG(SSP, 2, 0, SSP2-Codec,
+	SSP_CONFIG(I2S, SSP_CLOCK(mclk, 19200000, codec_mclk_in),
+		SSP_CLOCK(bclk, 3072000, codec_slave),
+		SSP_CLOCK(fsync, 48000, codec_slave),
+		SSP_TDM(2, 32, 3, 3),
+		SSP_CONFIG_DATA(SSP, 2, 16, 1, 0, 10)))
+
+# dmic01 (id: 1)
+DAI_CONFIG(DMIC, 0, 1, dmic01,
+	DMIC_CONFIG(1, 2400000, 4800000, 40, 60, 48000,
+		DMIC_WORD_LENGTH(s32le), 400, DMIC, 0,
+		PDM_CONFIG(DMIC, 0, FOUR_CH_PDM0_PDM1)))
+
+# 3 HDMI/DP outputs (ID: 2,3,4)
+DAI_CONFIG(HDA, 3, 2, iDisp1,
+	HDA_CONFIG(HDA_CONFIG_DATA(HDA, 3, 48000, 2)))
+DAI_CONFIG(HDA, 4, 3, iDisp2,
+	HDA_CONFIG(HDA_CONFIG_DATA(HDA, 4, 48000, 2)))
+DAI_CONFIG(HDA, 5, 4, iDisp3,
+	HDA_CONFIG(HDA_CONFIG_DATA(HDA, 5, 48000, 2)))
+
+#SSP 1 (ID: 5) with 19.2 MHz mclk with MCLK_ID 1 (unused), 1.536 MHz blck
+DAI_CONFIG(SSP, 1, 5, SSP1-Codec,
+	SSP_CONFIG(I2S, SSP_CLOCK(mclk, 19200000, codec_mclk_in),
+		SSP_CLOCK(bclk, 1536000, codec_slave),
+		SSP_CLOCK(fsync, 48000, codec_slave),
+		SSP_TDM(2, 16, 3, 3),
+		SSP_CONFIG_DATA(SSP, 1, 16, 1)))
```

Comment:
this is too complicated, there is no reason to reorder the BE dailinks, we can change the machine driver and first add the two SSP links, then DMIC and last HDMI. that way only the SSP number and configuration will change, and dmic/HDMI can be moved out of the conditional part.

Reply:
got it. will update the commits.